### DISCUSSION TITLE
Join the Conversation: Wunjo AI Community Forum

### DISCUSSION THREAD

**Discussion by wladradchenko** [wladradchenko/wunjo.wladradchenko.ru#2] [Announcements]
## 👋 Welcome!
  **First of all**, I want to tell you about the [Community channel](https://boosty.to/wunjo), where Wunjo news, voting for functionality, tutorials, and early access appear. Join to channel if you care about the fate of the project.
  We’re using **Discussions** as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by wladradchenko** [comment]
I will be glad to answer your questions

**Reply by realyussel** [reply]
Hello, your project is great, however I speak Spanish and I would like to know how to create a voice cloning synthesis model for my language. Even if you could give me some direction, I would like to work with you to integrate Spanish into the available languages.

**Reply by wladradchenko** [reply]
Hello,

Thank you for your interest in Wunjo project and for your willingness to collaborate in expanding its capabilities to include the Spanish language. This is sound cool.

### Step 1: Repository and Initial Setup

Firstly, I use the repository available at [Real-Time-Voice-Cloning GitHub](https://github.com/CorentinJ/Real-Time-Voice-Cloning/wiki/Training) for cloning voices. Here, you only need to train the `synthesizer.pt` as the default encoder and vocoder can be utilized as is. These default models will be automatically generated in the `saved_models` directory during the first run.

### Step 2: Dataset Preparation

Finding a high-quality Spanish dataset is pivotal. Although I'm not certain of where to obtain a Spanish dataset, it should have a structure similar to that of LibriTTS:

```
datasets_root
     └─LibriTTS
         └─train-clean-100
             └─speaker-001
                 └─book-001
                     ├─utterance-001.wav
                     ├─utterance-001.txt
                     ├─utterance-002.wav
                     └─utterance-002.txt
```

![How view book-001 for Russian](https://github.com/wladradchenko/wunjo.wladradchenko.ru/assets/56233697/ab63cf29-b757-4c13-918e-b7e8aeda3bff)

It means what you can have book-001, bool-002, book-003, etc. in speaker-001. Also you can have speaker-001, speaker-002, speaker-003, etc. in train-clean-100. Name of audio equivalent to name of txt file, in txt file will be text from audio.

### Step 3: Adapting for the Spanish Language

To tailor the system for the Spanish language, adjust the symbols in the [symbols.py file](https://github.com/CorentinJ/Real-Time-Voice-Cloning/blob/master/synthesizer/utils/symbols.py) to include Spanish letters.


### Step 4: Data Preprocessing

Once you have your dataset, preprocess the audio files using the following commands:

```
python synthesizer_preprocess_audio.py datasets_root --datasets_name LibriTTS --subfolders train-clean-100 --no_alignments
python synthesizer_preprocess_embeds.py datasets_root/SV2TTS/synthesizer/
```

### Step 5: Training the Model

Start the training with the following command:

```
python synthesizer_train.py spanish datasets_root/SV2TTS/synthesizer
```

To expedite the training process, modify the batch size in the [hparams.py file](https://github.com/CorentinJ/Real-Time-Voice-Cloning/blob/master/synthesizer/hparams.py):

```
tts_schedule = [
    (2, 1e-3, 20_000, 32),
    (2, 5e-4, 40_000, 32),
    (2, 2e-4, 80_000, 32),
    (2, 1e-4, 160_000, 32),
    (2, 3e-5, 320_000, 32),
    (2, 1e-5, 640_000, 32)
]
```

and set tts_cleaner_names as basic_cleaners to not miss specific languages letter without ASCII translate. 

```
tts_cleaner_names = ["basic_cleaners"]
```

![How you training voice cloning russian model](https://github.com/wladradchenko/wunjo.wladradchenko.ru/assets/56233697/ba9e8de0-6241-4474-9fc3-c837ca5314fb)

P.S. It is important to note that a larger number of iterations does not necessarily imply a better-performing model. For instance, during my experience training a model to clone a Russian voice, the optimum quality was achieved at the 175,000th iteration. Surprisingly, at 200,000 and even beyond 250,000 iterations, the model began to generate a robotic voice. I speculate that this anomaly might be related to the dataset I used, which is available at [open_stt](https://github.com/snakers4/open_stt). Upon closer examination of the dataset, I noticed it contains audio files with artificial Russian voices, which could potentially influence the training outcomes. In [test.zip](https://github.com/wladradchenko/wunjo.wladradchenko.ru/files/12705757/test.zip) you will able to listen how Russian cloning voice sound and difference iteration. 

When the training process will finish you would find Spanish synthesizer.pt in `synthesizer/saved_models/spanish`. You can test the quality of the trained model by running `python demo_cli.py -s synthesizer_path_model/synthesizer.pt`.

And in order for me to implement voice cloning for the Spanish Language in Wunjo AI, I only need symbols.py file including Spanish letters and the synthesizer.pt model.

**Reply by wladradchenko** [reply]
Also, I asked ChatGPT about Spanish:

### Spanish Linguistic Characteristics

#### 1. Diacritical Marks
Spanish uses several diacritical marks, including:
   - **Acute accents (´)**: used to mark stress or to distinguish between homonyms, e.g., "sí" (yes) vs. "si" (if).
   - **Tilde (~)**: used in "ñ" to represent a palatal nasal sound. (but in cloning voice I use ~ to split phrase on small phrases, this is not problem to change)
   - **Dieresis (¨)**: used in "ü" to indicate that "u" should be pronounced in the combinations "güe" and "güi".
   
You might want to add modules that correctly handle words with these marks, ensuring that they are correctly pronounced.

P.S. However, unlike Russian, in Spanish it is clear that there is something above the letter, so it seems to me that there is no need to do anything for this. Just save the original letter, and the neural network will understand that it needs to be emphasized, etc. The symbols are enough for this.

#### 2. Homographs
Similar to Russian, Spanish has homographs (words spelled the same but pronounced differently), e.g., "te" (tea) and "te" (you - object pronoun). You might consider adding a module to correctly emphasize these words based on the context.

#### 3. Stress
Word stress is significant in Spanish. Most words follow regular stress rules, but there are exceptions. You might add a module that correctly identifies the stressed syllable in each word, including those that don't follow the standard rules.

### Modules for Spanish
Given the linguistic characteristics outlined above, here are some modules you might consider for Spanish:

1. **Acute Accent Handler**: To correctly handle words with acute accents. (I think enough keep ´ upper character)
2. **Tilde Handler**: To correctly handle words containing the "ñ" character. (I think enough keep ~ upper character)
3. **Dieresis Handler**: To manage the correct pronunciation of "ü" in various contexts. (I think enough keep ¨ upper character)
4. **Homograph Handler**: To differentiate between homographs based on context.
5. **Stress Handler**: To identify and correctly pronounce the stressed syllable in each word.

You can see witch I use Russian files in [folder](https://github.com/wladradchenko/wunjo.wladradchenko.ru/tree/main/portable/src/speech/tps/rules). In order to Spanish speech will be more clear, need to modify the TPS module in Wunjo AI. I can do this for the Spanish language, however I just need these files for the Spanish words to I will use replace.

![Screenshot from 2023-09-21 08-45-32](https://github.com/wladradchenko/wunjo.wladradchenko.ru/assets/56233697/d4f8b0d0-d00a-4a93-a4b8-999d2840edae)

If add this in Wunjo AI, I think what will be available training speech model in interface Wunjo AI. I have recorded video about [this](https://youtu.be/KY7mdB9rXbw?feature=shared).

**Comment by omikron88** [comment]


**Reply by wladradchenko** [reply]
Hello,

For project launch and bug-related questions, please contact the owner of the original repository at https://github.com/CorentinJ/Real-Time-Voice-Cloning/wiki/Training.

Hovewer, to get started with the Real-Time-Voice-Cloning project, I recommend installing numpy==1.23.5 first for compatibility. You can do this using:

```
python -m pip install numpy==1.23.5
```

After that, proceed with the rest of the dependencies as per the project's documentation. Also you can try to install Python3.10 and use ones in that project.

**Comment by nine1one** [comment]
Hi,

When we run the instance on PC, can it be used locally on phones??

**Reply by wladradchenko** [reply]
Hi. It is flask app. U can install the application on server as web app and share to your smartphone in browser. But I didn't the frontend for mobile device, this css you need to adaptive independently for your mobile or tablet resolution.

If u will do CSS to adaptive in mobile size you can pull your changes in project to this will available for other users.

Also, I wanna note what backend of application work on post and get request. And frontend can be created on any technology and language for example on React Native or Kotlin, etc.

**Reply by nine1one** [reply]
https://github.com/ShiftHackZ/Stable-Diffusion-Android

This is where I got the idea to ask you.

And there is one here that I am learning from..

https://github.com/ZTMIDGO/Android-Stable-diffusion-ONNX

**Reply by wladradchenko** [reply]
In Wunjo AI a lot neural network models witch can not be adaptive to run on smartphone and requires a VRAM and powerful processor. But u can use backend from Wunjo AI on linux server and mobile frontend witch will work with post and get requests. 

For now, I don’t plan to make a mobile application, because I haven’t fully realized my vision in the desktop version yet.

**Reply by nine1one** [reply]
I see and my best wishes are with you and your app. Wunjo in my opinion is so far on of the best app and it's very under-rated. I hope it gets the hype it deserves soon. I have been trying to build the Windows GPU version but I haven't got success yet. It would be more than great if you can check the steps once and update if there is any changes required. I have been able to fix some of the mismatches but still not able to build a result. I am a newbie :))

**Reply by wladradchenko** [reply]
Thank you for your kind words and support! It's truly motivating to hear that Wunjo has user like you. If you're enjoying the app, sharing your experience on platforms like Reddit, YouTube, TikTok or any platform can really help others discover it too, and I would be incredibly grateful for that.

Regarding the Windows GPU build, I'm curious to know if there's a specific reason the portable version from an official website isn't meeting your needs. This could help me understand and address any gaps in the current offerings.

When I built the portable version for Windows GPU I have recorded a video of display screen, so that I could later transfer the steps into the documentation, that video without comments and editing, since it is for personal use, but you can see all the steps.  If you think this uncut insight into the build process could assist you, I'd be happy to share it.

**Reply by nine1one** [reply]
> Thank you for your kind words and support! It's truly motivating to hear that Wunjo has user like you. If you're enjoying the app, sharing your experience on platforms like Reddit, YouTube, TikTok or any platform can really help others discover it too, and I would be incredibly grateful for that.
> 
> 
> 
> Regarding the Windows GPU build, I'm curious to know if there's a specific reason the portable version from an official website isn't meeting your needs. This could help me understand and address any gaps in the current offerings.
> 
> 
> 
> When I built the portable version for Windows GPU I have recorded a video of display screen, so that I could later transfer the steps into the documentation, that video without comments and editing, since it is for personal use, but you can see all the steps.  If you think this uncut insight into the build process could assist you, I'd be happy to share it.

Hi,

Yes and thank you. I noticed the new download info. I am sharing the app today on the platforms i am using. 

Thank you

**Reply by nine1one** [reply]
> Thank you for your kind words and support! It's truly motivating to hear that Wunjo has user like you. If you're enjoying the app, sharing your experience on platforms like Reddit, YouTube, TikTok or any platform can really help others discover it too, and I would be incredibly grateful for that.
> 
> 
> 
> Regarding the Windows GPU build, I'm curious to know if there's a specific reason the portable version from an official website isn't meeting your needs. This could help me understand and address any gaps in the current offerings.
> 
> 
> 
> When I built the portable version for Windows GPU I have recorded a video of display screen, so that I could later transfer the steps into the documentation, that video without comments and editing, since it is for personal use, but you can see all the steps.  If you think this uncut insight into the build process could assist you, I'd be happy to share it.

I am almost new to this, and I am still learning a lot from you. I cannot express my gratitude towards you and your project. It's helping me learn a lot like I recently burned my Ryzen 7cpu. 😂

**Comment by omikron88** [comment]
Is this useable to implement Czech:
[https://github.com/comodoro/deepspeech-cs](https://github.com/comodoro/deepspeech-cs)

**Reply by wladradchenko** [reply]
Hi, I opened a discussion for this by [link](https://github.com/wladradchenko/wunjo.wladradchenko.ru/discussions/31)

**Comment by JunYu1996** [comment]
For some reason, wunjo cannot be started using the easy installation method and manual transfer method, and it will automatically open a page with 127.0.0.1 @wladradchenko

**Reply by wladradchenko** [reply]
Hello. This is flask app, of course this will start in localhost and browser. It is also worth noting that launch problems are not a topic for discussion, but issues. And in order to understand issue beed to compare more information. 

Anyway, you can visit documentation page or watch [video](https://youtube.com/playlist?list=PLJG0sD6007zFJyV78mkU-KW2UxbirgTGr&feature=shared) how install Wunjo from code or installation. There are absolutely all the answers to any problems that users have had.

You may have a launch problem because you launched more than once and did not close previous sessions.  To close the session, close your browser, or use a special browser that I made for the application.  With this browser there will be no problems with re-enabling.

**Comment by JuliaNavag** [comment]
Hi, I try to open wunjo comunity edition but it crash everytime.
This is the log the app show:


Traceback (most recent call last):
  File "\app\insightface\__init__.py", line 8, in <module>
    import onnxruntime
  File "\app_packages\onnxruntime\__init__.py", line 57, in <module>
    raise import_capi_exception
  File "\app_packages\onnxruntime\__init__.py", line 23, in <module>
    from onnxruntime.capi._pybind_state import ExecutionMode  # noqa: F401
  File "\app_packages\onnxruntime\capi\_pybind_state.py", line 32, in <module>
    from .onnxruntime_pybind11_state import *  # noqa
ImportError: DLL load failed while importing onnxruntime_pybind11_state: Error en una rutina de inicialización de biblioteca de vínculos dinámicos (DLL).

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "\app\wunjo\__main__.py", line 13, in <module>
    from wunjo.app import main
  File "\app\wunjo\app.py", line 24, in <module>
    from deepfake.inference import FaceSwap, Enhancement, SAMGetSegment, RemoveBackground, RemoveObject, LipSync, Analyser
  File "\app\deepfake\inference.py", line 25, in <module>
    from src.utils.sync_lip import GenerateWave2Lip
  File "\app\deepfake\src\utils\sync_lip.py", line 12, in <module>
    from src.face3d.recognition import FaceRecognition
  File "\app\deepfake\src\face3d\recognition.py", line 6, in <module>
    import insightface
  File "\app\insightface\__init__.py", line 10, in <module>
    raise ImportError(
ImportError: Unable to import dependency onnxruntime.

**Reply by wladradchenko** [reply]
Hello! GitHub Discussions is generally used for community questions and sharing knowledge, so for reporting bugs, it might be more effective to open an issue in the repository’s Issues section.

As for installation help, did you have a chance to check out the installation video? It provides step-by-step guidance, including installing Visual Studio, which is required for setting up the development environment. Once Visual Studio is installed, you’ll need to follow the instructions in the video carefully to reinstall the dependencies correctly. Here’s the link to the video for easy reference: [Installation Video](https://www.youtube.com/watch?v=00gzOC_cw9Y&list=PLJG0sD6007zFJyV78mkU-KW2UxbirgTGr&index=7) and console command in [Wiki](https://github.com/wladradchenko/wunjo.wladradchenko.ru/wiki/How-to-use-the-GPU-in-the-application).

**Reply by JuliaNavag** [reply]
Hi, thanks for the advice about post.

About the problems, it is not in the installation process, the installation process goes perfect, I use the installation file from https://wunjo.online/,  this problem ocurs when I open the installed program.

Thanks again.

**Comment by surfmore** [comment]
I would like to help with Brazilian Portuguese launguage, if nobody is taking care of it already.

**Reply by wladradchenko** [reply]
Hi. I've been to Brazil, in my humble opinion, it's one of the best countries. If you want, to support you can help with your vote on [DevHunt](https://devhunt.org/tool/wunjo-community-edition-ce) - it motivates me to add new features to the open source version of the application, including localization into other languages. Currently, the interface only supports English.

**Reply by surfmore** [reply]
I meant on tts or stt features. You said you needed help somewhere to train in other languages

**Reply by wladradchenko** [reply]
> I meant on tts or stt features. You said you needed help somewhere to train in other languages

It was for version 1.x. Last version 2.x of Wunjo Pro has clone voice on any language instead using tts and has auto translation interface. Wunjo Community Edition is not support clone voice and I don't have plans to update CE on clone voice before 4096 stars.

**Comment by therealbappy** [comment]
Hello ,

Is it possible to use other safetensor for generation module ? 

i would like to use this one https://civitai.com/models/978314/ultrareal-fine-tune

have you got a tutorial to use custom model  ? i checked on your video but don t find anyone.

Best regards,

**Reply by wladradchenko** [reply]
Hi. I tried to implement Flux to improve quality, including with micro learning, to generate videos with the same object. But for now, I need more than 8 GB of video memory, and this is more than I have on my device. I am following several projects, how they adapt Flux (with Inpaint, with Controlnet) for 8 GB (ComfyUI does not count, I do not understand what optimizations are there). So, Flux will appear in the future, but for now it is not there. At the moment, you can use video generation from an image, including creating images, using Stable diffusion 1.5.

**Reply by therealbappy** [reply]
Sorry, i didn t check that m'y link was flux.

So for exemple for a stable diffusion model mike this one : https://civitai.com/models/153568/real-dream?modelVersionId=712448

Can i use it in wunjo pro local version ?
If yes , how Can i configure wunjo to install it like other model,must i have to édit diffusion.json ?

**Reply by wladradchenko** [reply]
@therealbappy Setting up custom_diffusion.json
In the file `.cache/wunjo/all_models/custom_diffusion.json` you specify the paths to your models. Example of setting up:


```
[
	{
		"name": "Fantasy World", "model": "fantasyWorld_v10.safetensors", "img": "https://image.civitai.com/xG1nkqKTMzGDvpLrqFT7WA/2e321e71-c144-4ba6-8b55-3f62756fc0a1/width=1024,quality=90/01828-3669354034-giant%20pink%20balloon,cities%20floating%20on%20balloons,magnificent%20clouds,pink%20theme,.jpeg", "url": ""
	}
]
```

You need to put your model in `.cache/wunjo/all_models/diffusion/your_model.safetensor` and write this name in "model" key of custom_diffusion.json.

If you specify the url, then the model is downloaded automatically. 

_To show that the url can be used for automatic download, when you paste it into your browser, the model will automatically start downloading without displaying the CivitAI page._

P.S. You can open .cache/wunjo folder from software, you need to click on Workspace title, and after click where wrote number of Gb models and `(click)` message.

For Real Dream model will be:
```
[
{
"name": "Real Dream", "model": "realDream_15SD15.safetensors", "img": "https://image.civitai.com/xG1nkqKTMzGDvpLrqFT7WA/148d35a7-fd98-4e53-9139-2dc1fd651d39/original=true,quality=90/tmp405itv8v.jpeg", "url": "https://civitai.com/api/download/models/712448?type=Model&format=SafeTensor&size=pruned&fp=fp16"
}
]
```

**Reply by therealbappy** [reply]
tx , it s working.

**Comment by therealbappy** [comment]
Hello ,

i got an issue on the new version Wunjo-2.0.6.

On the login , in the password textbox , we cannot put special caractere like "!,?$" ...

**Reply by wladradchenko** [reply]
Hi. Yes, I have set such a restriction. If the password already contains such characters, use the password change. So, this is not issue, this is feature 😅

**Reply by therealbappy** [reply]
ok , i think it s less secure but ok

**Comment by krtsimmons** [comment]
can i use your project for real time voice call?

**Reply by wladradchenko** [reply]
Real-time voice cloning is not supported.
